app/models/category_list.rb

```diff
@@ -135,21 +135,21 @@ def dismissed_topic?(topic)
   end
 
   def find_categories
+    # Enforce paginaion for users who can see a large number of categories to
+    # smooth out the performance of the category list page.
+    paginate =
+      Category.secured(@guardian).count > MAX_UNOPTIMIZED_CATEGORIES ||
+        @guardian.can_lazy_load_categories?
+
     query = Category.includes(CategoryList.included_associations).secured(@guardian)
     query = self.class.order_categories(query)
 
-    if @options[:parent_category_id].present? || @guardian.can_lazy_load_categories?
+    if @options[:parent_category_id].present? || paginate
       query = query.where(parent_category_id: @options[:parent_category_id])
     end
 
-    style =
-      if Category.secured(@guardian).count > MAX_UNOPTIMIZED_CATEGORIES
-        "categories_only_optimized"
-      else
-        SiteSetting.desktop_category_page_style
-      end
```

Comment:
I think this conditional is still valid? If a user can see a ton of categories, we want to force the `categories_only` style cause paginating on the other styles is kind of broken UX wise.